File: native/aloud-tts/android/src/main/java/com/aloud/tts/AloudCore.kt
package com.aloud.tts

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import org.json.JSONObject

/**
 * Kotlin binding to the `aloud_core` C ABI via JNA (the same mechanism `uniffi`
 * uses). This is the Android end of the contract in
 * `contracts/ffi.contract.md`; it owns the session pointer and guarantees the
 * dispatch result string is freed.
 *
 * The shared library `libaloud_core.so` is produced by `cargo-ndk` per ABI and
 * packaged under `jniLibs/` (see android/README.md).
 */
private interface AloudCoreLib : Library {
    fun aloud_core_version(): Pointer
    fun aloud_session_new(text: String): Pointer?
    fun aloud_session_free(session: Pointer?)
    /**
     * The C ABI declares `uint32_t`, so this MUST be a 32-bit `Int` — JNA maps
     * Kotlin `Long` to a 64-bit native integer unconditionally. It previously
     * declared `Long` against a `size_t` return, which is 4 bytes on
     * armeabi-v7a: JNA read 8, so the high half was whatever happened to be in
     * the register. Correct on arm64, garbage on 32-bit. See the "fixed-width"
     * rule in contracts/ffi.contract.md.
     */
    fun aloud_session_unit_count(session: Pointer?): Int
    fun aloud_session_dispatch(session: Pointer?, commandJson: String): Pointer
    fun aloud_string_free(ptr: Pointer?)
}

/** A parse error surfaced from the core as a typed exception. */
class CoreException(val code: String, override val message: String) : Exception(message)

data class Highlight(val start: Int, val end: Int)

data class Snapshot(
    val status: String,
    val unit: Int,
    val unitCount: Int,
    val token: Int,
    val tokenCount: Int,
    val utterance: String,
    val highlight: Highlight?,
)

/**
 * Owns one Rust session pointer.
 *
 * ## Ownership rule (the same one [AloudCore.swift] implements)
 * The session is freed **exactly once**. Swift gets that for free from ARC via
 * `deinit`; Kotlin has no deterministic destructor, so the caller must call
 * [release] — but calling it more than once, or dispatching after it, must never
 * corrupt memory. `aloud_core.h` is explicit that double-freeing is undefined
 * behaviour, so that guarantee lives here, in the type that owns the pointer,
 * rather than in every caller remembering to null its reference.
 *
 * Known limitation: there is no finalizer, so an instance dropped **without**
 * [release] leaks the Rust session until the process exits. That is the right
 * trade for a JNA binding — finalizers are deprecated and unreliable, and a
 * leak is far less harmful than a free racing the GC — but it does mean
 * `release()` is mandatory, not merely polite. [AloudTtsModule] calls it from
 * both `release()` and `load()`.
 */
class AloudCore(text: String) {
    private var session: Pointer? =
        LIB.aloud_session_new(text) ?: throw CoreException("INVALID_UTF8", "core rejected document text")

    /** The live pointer, or a typed error if the session was already released. */
    private fun requireSession(): Pointer =
        session ?: throw CoreException("NULL_POINTER", "core session has been released")

    val unitCount: Int get() = LIB.aloud_session_unit_count(requireSession())

    /** Send a command object and decode the response into a [Snapshot]. */
    fun dispatch(command: JSONObject): Snapshot {
        val resultPtr = LIB.aloud_session_dispatch(requireSession(), command.toString())
        try {
            val json = resultPtr.getString(0, "UTF-8")
            val obj = JSONObject(json)
            if (obj.has("error")) {
                val err = obj.getJSONObject("error")
                throw CoreException(err.getString("code"), err.getString("message"))
            }
            return obj.toSnapshot()
        } finally {
            // The core owns the buffer until we hand it back.
            LIB.aloud_string_free(resultPtr)
        }
    }

    /**
     * Release the Rust session. **Idempotent** — calling it again is a no-op,
     * and any later use throws a typed [CoreException] instead of dereferencing
     * a dangling pointer. Callers still should not use the core afterwards; this
     * only guarantees that doing so fails loudly rather than corrupting memory.
     */
    fun release() {
        // Clear the field first, so a concurrent or re-entrant call cannot pass
        // the same pointer to the allocator twice.
        val handle = session ?: return
        session = null
        LIB.aloud_session_free(handle)
    }

    companion object {
        /** Convenience command builders, matching the Rust `#[serde(tag="type")]`. */
        fun play() = JSONObject().put("type", "Play")
        fun pause() = JSONObject().put("type", "Pause")
        fun next() = JSONObject().put("type", "Next")
        fun prev() = JSONObject().put("type", "Prev")
        fun getState() = JSONObject().put("type", "GetState")
        fun seekUnit(unit: Int) = JSONObject().put("type", "SeekUnit").put("unit", unit)
        fun seekByte(byte: Int) = JSONObject().put("type", "SeekByte").put("byte", byte)
        fun wordBoundary(utf16Offset: Int) =
            JSONObject().put("type", "WordBoundary").put("utf16Offset", utf16Offset)

        val version: String get() = LIB.aloud_core_version().getString(0, "UTF-8")

        private val LIB: AloudCoreLib by lazy {
            Native.load("aloud_core", AloudCoreLib::class.java)
        }
    }
}

private fun JSONObject.toSnapshot(): Snapshot {
    val highlight = optJSONObject("highlight")?.let {
        Highlight(it.getInt("start"), it.getInt("end"))
    }
    return Snapshot(
        status = getString("status"),
        unit = getInt("unit"),
        unitCount = getInt("unitCount"),
        token = getInt("token"),
        tokenCount = getInt("tokenCount"),
        utterance = getString("utterance"),
        highlight = highlight,
    )
}

File: native/aloud-tts/ios/AloudCore.swift
import AloudCoreFFI // the Rust core's C ABI, exposed as a Clang module (see core/include/module.modulemap)
import Foundation

/// A safe Swift wrapper around the `aloud_core` C ABI (see
/// `core/include/aloud_core.h`). Nothing else in the iOS layer touches the raw
/// pointers: this type owns the session handle and guarantees it is freed, and
/// it translates the JSON dispatch protocol into typed `Snapshot`/`CoreError`.
///
/// This is the Swift end of the contract in `contracts/ffi.contract.md`.
///
/// ## Ownership rule (the same one AloudCore.kt implements)
/// The session is freed **exactly once**. Here that falls out of ARC: `session`
/// is a `let`, so it can never be reassigned, and `deinit` runs once when the
/// last reference goes away. The Kotlin binding has no deterministic destructor,
/// so it exposes an idempotent `release()` instead — different mechanics, same
/// guarantee, because `aloud_core.h` states that double-freeing is undefined
/// behaviour.
final class AloudCore {
    /// Opaque session pointer from `aloud_session_new`. `let`, so ARC's `deinit`
    /// is the single path to `aloud_session_free`.
    private let session: OpaquePointer

    /// Engine version, surfaced for diagnostics so JS/native/Rust logs agree.
    static var version: String {
        String(cString: aloud_core_version())
    }

    /// Parse `text`. Throws if the core rejects it (e.g. allocation failure).
    init(text: String) throws {
        guard let handle = text.withCString({ aloud_session_new($0) }) else {
            throw CoreError(code: "INVALID_UTF8", message: "core rejected document text")
        }
        self.session = handle
    }

    deinit {
        aloud_session_free(session)
    }

    /// The C ABI returns a fixed-width `uint32_t` (see the "fixed-width" rule in
    /// contracts/ffi.contract.md); widen it to Swift's native `Int` at the edge.
    var unitCount: Int {
        Int(aloud_session_unit_count(session))
    }

    /// Send a command and decode the response. A single choke point keeps the
    /// UTF-8/JSON handling — and the memory ownership of the returned string —
    /// in exactly one place.
    func dispatch(_ command: Command) throws -> Snapshot {
        let json = try JSONEncoder().encode(command)
        let raw = String(decoding: json, as: UTF8.self)

        guard let responsePtr = raw.withCString({ aloud_session_dispatch(session, $0) }) else {
            throw CoreError(code: "NULL_POINTER", message: "dispatch returned null")
        }
        // The core owns the returned buffer until we free it; copy out first.
        defer { aloud_string_free(responsePtr) }
        let data = Data(bytes: responsePtr, count: strlen(responsePtr))

        // The response is either a Snapshot or an error envelope.
        if let envelope = try? JSONDecoder().decode(ErrorEnvelope.self, from: data) {
            throw envelope.error
        }
        return try JSONDecoder().decode(Snapshot.self, from: data)
    }
}
